PR: https://github.com/odoo/odoo/pull/256889

From: f1f5a0b8f8a3570b0c72072cdf194c5a6bb2c3d3
From: David Monnom (moda)
Date: 2026-07-18 12:14:23

Breaking data model changes scores: del:2 + add:0, change matches:
-    is_online_payment = fields.Boolean(string="Online Payment", help="Use this payment method for online payments (payments made on a web page with online payment providers)", default=False)
-    _inherit = 'pos.session'

Total Changes: 230

[REF] point_of_sale: accounting

Refactor of the pos accounting and invoicing logic, now when closing
a session, it will create two moves if needed
* Sales move: containing all pos.order of type sale with aggregated
lines
* Refunds move: containing all pos.order of type refund with aggregated
lines

One of the major changes here is `pos_settle_due` in which is now only
possible to settle invoices and not pos order. That's because when
paying by customer account it will create an invoice each time.

Removed fields
* `invoice_journal_id` and `pos_invoice_journal_id` because now all
invoice are created with the same journal, the one defined on the pos
config.

* `split_transactions` because now, we don't allow to split payments
transactions on cash and bank, only when the type is pay_later

* `account_default_pos_receivable_account_id` because now the receivable
account is defined though a default partner on the pos.config

* `cash_real_transaction`, `cash_register_balance_start`,
`cash_register_balance_end_real`, `cash_journal_id`,
`cash_register_difference` and `cash_register_balance_end` all these
infos are now handled though the cash statement and its lines, so no
need to have them on the session

* `is_cash_count` because cash is now identified by type == 'cash'

* `is_invoiced` replaced by `is_singly_invoiced` and
`is_globally_invoiced`

* `session_move_id` because account_move is now used to link the order
to the session move if it was not invoiced

* `is_online_payment` because online payment method is now identified by
type == 'online' and not by a boolean field

* `settle_invoice_product_id`, `deposit_product_id` and
`settle_due_product_id` because we don't create a pos.order when
settling an invoice, so no need to have these products.

* `settled_invoice_id` and `settled_order_id` because we don't create a
pos.order when settling an invoice, so no need to link them anymore.

* `init_customer_due_total`, `customer_due_total`,
`settled_order_line_ids`, `settled_orders_count`, `commercial_partner_id`
because we don't settle pos.order anymore

* `pos_order_line_ids` and `pos_amount_unsettled` because we don't settle
pos.order anymore

* `pos_orders_amount_due` because we don't settle pos.order anymore

Test tracking
A lot of tests were replaced because all accounting entries for order
and session closing are now refactored.

Some tests were also removed because they were not relevant anymore with
the new way of handling payment and invoicing on pos.order.

Replaced tests
`test_classic_sale_order` replace:
* `test_state_when_closing_register`

`test_fifo_valuation_with_invoice` replace:
* `test_fifo_valuation_with_invoice_when_pos_customer_is_delivery_type`

`test_fixed_tax_negative_qty_should_be_negative` replace:
* `test_fixed_tax_negative_qty`
* `test_fixed_tax_positive_qty`

`test_user_right_on_statement_line_for_pos_user` replace:
* `test_pos_closing_cash_details`

`test_rounding_when_closing_session` replace:
* `test_invoice_rounding_overpaid_backend`
* `test_pos_create_account_move_round_globally`
* `test_rounding_method`

Moved in `test_pos_accounting.py`:
`test_journal_entries_category_without_account`
`test_tax_change_blocked_when_open_pos_session`
`test_order_partial_refund_rounding`

Test suite in `test_pos_accounting.py` replace:
* `test_pos_simple_invoiced_orders.py`
* `test_orders_no_invoiced`
* `test_orders_with_invoiced`
* `test_01_invoiced_order_from_other_customer`
* `test_02_all_orders_invoiced_mixed_customers`
* `test_total_due.py`
* `test_02_orders_without_invoice`
* `test_03_orders_with_invoice`

`test_invoice_a_negative_order_should_create_credit_note` replace:
* `test_manual_refund_negative_qty_invoice_creates_credit_note`

`test_invoice_an_order_from_closed_session` replace:
* `test_invoice_past_order_affecting_taxes`
* `test_invoice_past_refund`
* `test_invoice_past_order`
* `test_sale_order_postponed_invoicing`
* `test_order_invoiced_after_session_closed`
* `test_order_invoiced_customer_account_after_session_closed`
* `test_sale_order_postponed_invoicing`

`test_pos_order_sale_and_refund_with_taxes` replace:
* `test_orders_with_invoiced`
* `test_return_order`

`test_pos_order_sale_and_refund_with_taxes_invoiced` replace:
* `test_pos_create_correct_account_move`
* `test_return_order_invoiced`

`test_pos_order_sale_and_refund_with_taxes_not_invoiced` replace:
* `test_pos_create_correct_account_move`
* `test_return_order`

`test_invoiced_order_with_discount_sale_and_refund_with_tax` replace:
* `test_pos_create_correct_account_move_round_globally_discount`
* `test_pos_create_correct_account_move_round_globally_discount_real_use_case`

`test_tax_change_blocked_when_open_pos_session` replace:
* `test_tax_is_used_when_in_transactions`

`test_pos_payment_direction_and_account` replace:
* `test_pos_payment_direction_and_accounts`

`test_cash_statement_opening_and_closing_consistency` replace:
* `test_cash_register_if_no_order`
* `test_cash_payments_should_reflect_on_next_opening`

`test_invoicing_zero_amount_pos_order` replace:
* `test_orders_with_zero_valued_invoiced`

`test_order_partial_refund_rounding` replace:
* `test_refund_rounding_backend`
* `test_order_partial_refund`

`test_pos_config_with_other_currency_than_company` replace:
* `test_order_to_payment_currency`

`test_accounting_items_when_closing_with_bank_difference` replace:
* `test_payment_difference_accounting_items`
* `test_state_when_closing_register`

`test_pos_order_with_closing_storno` replace:
* `test_sale_order_postponed_invoicing_storno`

`test_various_orders` replace:
* `test_correct_partner_on_invoice_receivables`

`test_pos_order_session_closing_with_fp` replace:
* `test_01_no_invoice_fpos`
* `test_02_no_invoice_fpos_no_tax_dest`

`test_available_stock_order_no_invoice` replace:
* `test_orders_no_invoiced`
* `test_return_order`

`test_available_stock_order_with_invoice` replace:
* `test_orders_with_invoiced`
* `test_return_order`

`test_pos_order_with_company_branch` replace:
* `test_pos_branch_account`

Removed tests
`test_ZATCA_blocks_settle_due_and_sale_on_same_order` not usefull
settle due doesn't create pos.order anymore

`test_order_with_deleted_tax` removed because isn't possible to edit
taxes with open orders anymore.

`test_split_payment_linked_to_accounting_partner` split transaction is
now only available when using customer account (type == pay_layer)

`test_pos_hr_session_name_gap` removed not useful

`test_order_to_invoice_no_tax` not useful

`test_order_invoiced_customer_account_after_session_closed` not useful,
customer account payment are always invoiced but this test is almost
the same `test_invoice_an_order_from_closed_session`

`test_start_balance_with_two_pos` useless test, cash payment cannot be
shared.

`test_different_customer_invoices_settlement_same_session_reconciled_separately`
settling an invoice doesn't create `pos.order` anymore

`test_split_cash_payments` removed because split on customer account
is now available only when using `pay_later` payment method.

`test_receivable_account_reconciliation.py`, `test_pos_reconcile` no
more necessary, payment are now directly added to the account.move

`test_pos_settling_account_resets_on_payment_screen_unmount` removed
because now when leaving payment screen it will delete the order if its
a settling one.

`test_pos_settle_due_with_rounding` deleted, wtf we round on a money
deposit ?

`test_split_payment_linked_to_accounting_partner` split_transactions is
no more available with classic pm, only with pay_later.

`test_05_tax_base_amount` removed tax base amount are handled and tested
by accounting helpers

`test_no_kitchen_confirmation_for_deposit_money` removed because deposit
money doesn't create a pos.order anymore.

`test_order_with_deleted_tax` is now not possible to archive a tax used
by an open order.

`fastValidate` hoot deleted in pos_settle_due because isn't possible to
use fast validation with settle customer invoice anymore.

`test_double_syncing_same_order` removed, sync data with the same UUID
will update the same order, but since a payment_ids create command is in
the data, its normal that this payment is created two times.

`test_session_name_gap` removed, sequences are never rollbacked, so the
test is not correct.

closes odoo/odoo#256889

Related: odoo/enterprise#112634
Related: odoo/upgrade#10068
Signed-off-by: David Monnom (moda) <moda@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_online_payment/models/__init__.py
+++ b/addons/pos_online_payment/models/__init__.py
@@ -7,4 +7,3 @@ from . import account_payment
 from . import payment_transaction
 from . import pos_config
 from . import pos_order
-from . import pos_session

--- a/addons/pos_online_payment/models/payment_transaction.py
+++ b/addons/pos_online_payment/models/payment_transaction.py
@@ -36,7 +36,7 @@ class PaymentTransaction(models.Model):
                 if tools.float_compare(tx.amount, 0.0, precision_rounding=pos_order.currency_id.rounding) <= 0:
                     raise ValidationError(_('The payment transaction (%d) has a negative amount.', tx.id))
 
-                if not tx.payment_id: # the payment could already have been created by account_payment module
+                if not tx.payment_id:  # the payment could already have been created by account_payment module
                     tx._create_payment()
                 if not tx.payment_id:
                     raise ValidationError(_('The POS online payment (tx.id=%d) could not be saved correctly', tx.id))

--- a/addons/pos_online_payment/models/pos_config.py
+++ b/addons/pos_online_payment/models/pos_config.py
@@ -14,7 +14,7 @@ class PosConfig(models.Model):
         for config in self:
             opm_amount = 0
             for pm in config.payment_method_ids:
-                if pm.is_online_payment:
+                if pm.type == 'online':
                     opm_amount += 1
                     if opm_amount > 1:
                         raise ValidationError(_("A POS config cannot have more than one online payment method."))
@@ -23,7 +23,7 @@ class PosConfig(models.Model):
 
     def _get_cashier_online_payment_method(self):
         self.ensure_one()
-        return self.payment_method_ids.filtered('is_online_payment')[:1]
+        return self.payment_method_ids.filtered(lambda pm: pm.type == 'online')[:1]
 
     @api.model
     def _create_online_payment_demo(self):

--- a/addons/pos_online_payment/models/pos_order.py
+++ b/addons/pos_online_payment/models/pos_order.py
@@ -1,14 +1,13 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import models, fields, api, tools
+from odoo import api, fields, models, tools
 
 
 class PosOrder(models.Model):
     _inherit = 'pos.order'
 
     online_payment_method_id = fields.Many2one('pos.payment.method', compute="_compute_online_payment_method_id")
-    next_online_payment_amount = fields.Float(string='Next online payment amount to pay', digits=0, required=False) # unlimited precision
+    next_online_payment_amount = fields.Float(string='Next online payment amount to pay', digits=0, required=False)  # unlimited precision
 
     @api.depends('config_id.payment_method_ids')
     def _compute_online_payment_method_id(self):
@@ -24,7 +23,7 @@ class PosOrder(models.Model):
 
     def _clean_payment_lines(self):
         self.ensure_one()
-        order_payments = self.env['pos.payment'].search(['&', ('pos_order_id', '=', self.id), ('online_account_payment_id', '=', False), ('payment_method_id.is_online_payment', '=', True)])
+        order_payments = self.env['pos.payment'].search(['&', ('pos_order_id', '=', self.id), ('online_account_payment_id', '=', False), ('payment_method_id.type', '=', 'online')])
         order_payments.unlink()
 
     def get_and_set_online_payments_data(self, next_online_payment_amount=False):
@@ -43,7 +42,7 @@ class PosOrder(models.Model):
         if is_paid:
             return {
                 'id': self.id,
-                'paid_order': self.read([], load=False)
+                'paid_order': self.read([], load=False),
             }
 
         online_payments = self.sudo().env['pos.payment'].search_read(domain=['&', ('pos_order_id', '=', self.id), ('online_account_payment_id', '!=', False)], fields=['payment_method_id', 'amount'], load=False)
@@ -54,7 +53,7 @@ class PosOrder(models.Model):
         }
         if not isinstance(next_online_payment_amount, bool):
             if tools.float_is_zero(next_online_payment_amount, precision_rounding=self.currency_id.rounding) and len(online_payments) == 0 and self.state == 'draft' and not self.config_id.module_pos_restaurant and len(self.config_id.trusted_config_ids) == 0:
-                self.sudo()._clean_payment_lines() # Needed to delete the order
+                self.sudo()._clean_payment_lines()  # Needed to delete the order
                 return_data['deleted'] = True
             elif self._check_next_online_payment_amount(next_online_payment_amount):
                 self.next_online_payment_amount = next_online_payment_amount
@@ -74,7 +73,7 @@ class PosOrder(models.Model):
     def _process_order(self, order, existing_order):
         draft = order.get('state') == 'draft'
         pos_session = self.env['pos.session'].browse(order['session_id'])
-        online_payment_methods = pos_session.config_id.payment_method_ids.filtered('is_online_payment')
+        online_payment_methods = pos_session.config_id.payment_method_ids.filtered(lambda pm: pm.type == 'online')
         if draft and online_payment_methods:
             # online payment lines should not be created in draft orders
             order['payment_ids'] = [payment for payment in order.get('payment_ids', []) if payment[0] not in [0, 1] or (payment[2].get('payment_method_id') not in online_payment_methods.ids)]

--- a/addons/pos_online_payment/models/pos_payment.py
+++ b/addons/pos_online_payment/models/pos_payment.py
@@ -1,8 +1,8 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import logging
-from odoo import api, fields, models, _
+
+from odoo import _, api, fields, models
 from odoo.exceptions import UserError
 
 _logger = logging.getLogger(__name__)
@@ -11,7 +11,7 @@ _logger = logging.getLogger(__name__)
 class PosPayment(models.Model):
     _inherit = 'pos.payment'
 
-    online_account_payment_id = fields.Many2one('account.payment', string='Online accounting payment', readonly=True) # One2one
+    online_account_payment_id = fields.Many2one('account.payment', string='Online accounting payment', readonly=True)  # One2one
 
     @api.model_create_multi
     def create(self, vals_list):
@@ -22,7 +22,7 @@ class PosPayment(models.Model):
                 online_account_payments_by_pm[pm_id] = set()
             online_account_payments_by_pm[pm_id].add(vals.get('online_account_payment_id') or None)
 
-        opms_read_id = self.env['pos.payment.method'].search_read(['&', ('id', 'in', list(online_account_payments_by_pm.keys())), ('is_online_payment', '=', True)], ["id"])
+        opms_read_id = self.env['pos.payment.method'].search_read(['&', ('id', 'in', list(online_account_payments_by_pm.keys())), ('type', '=', 'online')], ["id"])
         opms_id = {opm_read_id['id'] for opm_read_id in opms_read_id}
         online_account_payments_to_check_id = set()
 
@@ -30,8 +30,7 @@ class PosPayment(models.Model):
             if pm_id in opms_id:
                 if None in oaps_id:
                     raise UserError(_("Cannot create a POS online payment without an accounting payment."))
-                else:
-                    online_account_payments_to_check_id.update(oaps_id)
+                online_account_payments_to_check_id.update(oaps_id)
             elif any(oaps_id):
                 raise UserError(_("Cannot create a POS payment with a not online payment method and an online accounting payment."))
 
@@ -43,13 +42,13 @@ class PosPayment(models.Model):
         return super().create(vals_list)
 
     def write(self, vals):
-        if vals.keys() & ('amount', 'payment_date', 'payment_method_id', 'online_account_payment_id', 'pos_order_id') and any(payment.online_account_payment_id or payment.payment_method_id.is_online_payment for payment in self):
+        if vals.keys() & ('amount', 'payment_date', 'payment_method_id', 'online_account_payment_id', 'pos_order_id') and any(payment.online_account_payment_id or payment.payment_method_id.type == 'online' for payment in self):
             raise UserError(_("Cannot edit a POS online payment essential data."))
         return super().write(vals)
 
     @api.constrains('payment_method_id')
     def _check_payment_method_id(self):
-        bypass_check_payments = self.filtered('payment_method_id.is_online_payment')
+        bypass_check_payments = self.filtered(lambda p: p.payment_method_id.type == 'online')
         if any(payment.payment_method_id != payment.pos_order_id.online_payment_method_id for payment in bypass_check_payments):
             # An online payment must always be saved for the POS, even if the online payment method is no longer configured/allowed in the pos.config, because in any case it is saved by account_payment and payment modules.
             _logger.warning("Allow to save a POS online payment with an unexpected online payment method")

--- a/addons/pos_online_payment/models/pos_payment_method.py
+++ b/addons/pos_online_payment/models/pos_payment_method.py
@@ -1,30 +1,25 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models, _
+from odoo import Command, _, api, fields, models
 from odoo.exceptions import ValidationError
 
 
 class PosPaymentMethod(models.Model):
     _inherit = "pos.payment.method"
 
-    is_online_payment = fields.Boolean(string="Online Payment", help="Use this payment method for online payments (payments made on a web page with online payment providers)", default=False)
     online_payment_provider_ids = fields.Many2many('payment.provider', string="Allowed Providers", domain="[('is_published', '=', True)]")
     has_an_online_payment_provider = fields.Boolean(compute='_compute_has_an_online_payment_provider', readonly=True)
-    type = fields.Selection(selection_add=[('online', 'Online')])
-
-    @api.model
-    def _load_pos_data_fields(self, config):
-        params = super()._load_pos_data_fields(config)
-        params += ['is_online_payment']
-        return params
+    type = fields.Selection(
+        selection_add=[('online', 'Online')],
+        ondelete={"online": "cascade"},
+    )
 
     @api.model
     def _load_pos_data_read(self, records, config):
         read_records = super()._load_pos_data_read(records, config)
 
         for record in read_records:
-            if not record.get('is_online_payment'):
+            if record['type'] != 'online':
                 continue
 
             pm = self.env['pos.payment.method'].browse(record['id']).exists()
@@ -32,18 +27,10 @@ class PosPaymentMethod(models.Model):
             record['_customer_required'] = bool(set(providers) & set(self._get_customer_required_providers_code()))
         return read_records
 
-    @api.depends('is_online_payment')
-    def _compute_type(self):
-        opm = self.filtered('is_online_payment')
-        if opm:
-            opm.type = 'online'
-
-        super(PosPaymentMethod, self - opm)._compute_type()
-
     def _get_online_payment_providers(self, pos_config_id=False, error_if_invalid=True):
         self.ensure_one()
         providers_sudo = self.sudo().online_payment_provider_ids
-        if not providers_sudo: # Empty = all published providers
+        if not providers_sudo:  # Empty = all published providers
             providers_sudo = self.sudo().env['payment.provider'].search([('is_published', '=', True)])
 
         if not pos_config_id:
@@ -55,40 +42,39 @@ class PosPaymentMethod(models.Model):
             raise ValidationError(_("All payment providers configured for an online payment method must use the same currency as the Sales Journal, or the company currency if that is not set, of the POS config."))
         return valid_providers
 
-    @api.depends('is_online_payment', 'online_payment_provider_ids')
+    @api.depends('type', 'online_payment_provider_ids')
     def _compute_has_an_online_payment_provider(self):
         for pm in self:
-            if pm.is_online_payment:
+            if pm.type == 'online':
                 pm.has_an_online_payment_provider = bool(pm._get_online_payment_providers())
             else:
                 pm.has_an_online_payment_provider = False
 
-    @api.constrains('config_ids', 'is_online_payment')
+    @api.constrains('config_ids', 'type')
     def _check_pos_config_online_payment(self):
         """ Check that each POS config has at most one online payment method,"""
-        for pm in self.filtered('is_online_payment'):
+        for pm in self.filtered(lambda p: p.type == 'online'):
             for config in pm.config_ids:
-                other_online_pms = config.payment_method_ids.filtered(lambda other_pm: other_pm.is_online_payment and other_pm.id != pm.id)
+                other_online_pms = config.payment_method_ids.filtered(lambda other_pm: other_pm.type == 'online' and other_pm.id != pm.id)
                 if other_online_pms:
                     raise ValidationError(_("The %s already has one online payment.", config.name))
 
     def _is_write_forbidden(self, fields):
-        return super(PosPaymentMethod, self)._is_write_forbidden(fields - {'online_payment_provider_ids'})
+        return super()._is_write_forbidden(fields - {'online_payment_provider_ids'})
 
     @api.model_create_multi
     def create(self, vals_list):
         for vals in vals_list:
-            if vals.get('is_online_payment', False):
+            if vals.get('type', '') == 'online':
                 self._force_online_payment_values(vals)
         return super().create(vals_list)
 
     def write(self, vals):
-        if 'is_online_payment' in vals:
-            if vals['is_online_payment']:
-                self._force_online_payment_values(vals)
+        if 'type' in vals and vals['type'] == 'online':
+            self._force_online_payment_values(vals)
             return super().write(vals)
 
-        opm = self.filtered('is_online_payment')
+        opm = self.filtered(lambda p: p.type == 'online')
         not_opm = self - opm
 
         res = True
@@ -106,7 +92,7 @@ class PosPaymentMethod(models.Model):
         if 'type' in vals:
             vals['type'] = 'online'
 
-        disabled_fields_name = ('split_transactions', 'receivable_account_id', 'outstanding_account_id', 'journal_id', 'is_cash_count', 'payment_provider', 'qr_code_method')
+        disabled_fields_name = ('receivable_account_id', 'outstanding_account_id', 'journal_id', 'payment_provider', 'qr_code_method')
         if if_present:
             for name in disabled_fields_name:
                 if name in vals:
@@ -119,7 +105,7 @@ class PosPaymentMethod(models.Model):
             vals['payment_method_type'] = 'none'
 
     def _get_provider_selection(self):
-        if len(self) == 1 and self.is_online_payment:
+        if len(self) == 1 and self.type == 'online':
             return []
         return super()._get_provider_selection()
 
@@ -130,13 +116,13 @@ class PosPaymentMethod(models.Model):
             If there is not, create a new one for the company and return it without adding the pos.config to it.
         """
         # Parameters are ids instead of a pos.config record because this method can be called from a web controller or internally
-        payment_method_id = self.env['pos.payment.method'].search([('is_online_payment', '=', True), ('company_id', '=', company_id), ('config_ids', 'in', pos_config_id)], limit=1).exists()
+        payment_method_id = self.env['pos.payment.method'].search([('type', '=', 'online'), ('company_id', '=', company_id), ('config_ids', 'in', pos_config_id)], limit=1).exists()
         if not payment_method_id:
-            payment_method_id = self.env['pos.payment.method'].search([('is_online_payment', '=', True), ('company_id', '=', company_id)], limit=1).exists()
+            payment_method_id = self.env['pos.payment.method'].search([('type', '=', 'online'), ('company_id', '=', company_id)], limit=1).exists()
             if not payment_method_id:
                 payment_method_id = self.env['pos.payment.method'].create({
                     'name': _('Online Payment'),
-                    'is_online_payment': True,
+                    'type': 'online',
                     'company_id': company_id,
                     'sequence': 3,
                 })
@@ -148,13 +134,70 @@ class PosPaymentMethod(models.Model):
                     ))
         return payment_method_id
 
-    @api.onchange('is_online_payment')
-    def _onchange_is_online_payment(self):
+    @api.onchange('type')
+    def _onchange_type(self):
         """Reset method to hide widget `pos_payment_provider_cards` in form view."""
         self.payment_method_type = 'none'
 
-    def _is_online_payment(self):
-        return self.is_online_payment
-
     def _get_customer_required_providers_code(self):
         return ['aps', 'flutterwave']
+
+    def _create_online_payment_line_transfer(self, session):
+        """
+        Since online payments are created in another account receivable than
+        the PoS one, we need to create a transfer journal entry to link them.
+
+        The online payment creates:
+          - Debit: Payment provider receivable (destination_account_id)
+          - Credit: Outstanding account
+
+        We create a transfer entry:
+          - Debit: POS receivable
+          - Credit: Payment provider receivable
+
+        This allows reconciliation with the POS session invoice payment_term lines.
+        """
+        self.ensure_one()
+        pos_receivable = session._get_receivable_account()
+        online_orders = session.order_ids.filtered_domain([
+            ('payment_ids.payment_method_id', '=', self.id),
+        ])
+
+        total_amount = sum(pay.amount for pay in online_orders.payment_ids)
+        account_payment = online_orders.payment_ids.online_account_payment_id
+        ref = _(
+            "Transfer Online payment %(account_name)s => %(pos_receivable_name)s",
+            account_name=account_payment.destination_account_id.name,
+            pos_receivable_name=pos_receivable.name,
+        )
+        transfer_move = self.env["account.move"].sudo().create({
+            "move_type": "entry",
+            "journal_id": session.config_id.journal_id.id,
+            "ref": ref,
+            "line_ids": [
+                Command.create({
+                    "account_id": pos_receivable.id,
+                    "debit": 0,
+                    "credit": total_amount,
+                    "name": _("POS Receivable Transfer"),
+                    "partner_id": account_payment.partner_id.id,
+                }),
+                Command.create({
+                    "account_id": account_payment.destination_account_id.id,
+                    "debit": total_amount,
+                    "credit": 0,
+                    "name": _("Online Payment Transfer"),
+                    "partner_id": account_payment.partner_id.id,
+                }),
+            ],
+        })
+
+        transfer_move._post()
+        return transfer_move.line_ids.filtered(
+            lambda line: line.account_id == pos_receivable,
+        )
+
+    def _create_payment_line(self, session, amount, account=None, message=None, partner=None):
+        if self.type == 'online':
+            return self._create_online_payment_line_transfer(session)
+        return super()._create_payment_line(session, amount, account, message, partner)

--- a/addons/pos_online_payment/models/pos_session.py
+++ b/None
@@ -1,70 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from collections import defaultdict
-from odoo import models, tools, _
-from odoo.exceptions import UserError
-
-
-class PosSession(models.Model):
-    _inherit = 'pos.session'
-
-    def _accumulate_amounts(self, data):
-        data = super()._accumulate_amounts(data)
-        amounts = lambda: {'amount': 0.0, 'amount_converted': 0.0}
-
-        split_receivables_online = defaultdict(amounts)
-        currency_rounding = self.currency_id.rounding
-        for order in self._get_order_for_session_closing():
-            for payment in order.payment_ids:
-                amount = payment.amount
-                if tools.float_is_zero(amount, precision_rounding=currency_rounding):
-                    continue
-                date = payment.payment_date
-                payment_method = payment.payment_method_id
-                payment_type = payment_method.type
-
-                if payment_type == 'online':
-                    split_receivables_online[payment] = self._update_amounts(split_receivables_online[payment], {'amount': amount}, date)
-
-        data.update({'split_receivables_online': split_receivables_online,})
-        return data
-
-    def _create_bank_payment_moves(self, data):
-        data = super()._create_bank_payment_moves(data)
-
-        split_receivables_online = data.get('split_receivables_online')
-        MoveLine = data.get('MoveLine')
-
-        online_payment_to_receivable_lines = {}
-
-        for payment, amounts in split_receivables_online.items():
-            split_receivable_line = MoveLine.create(self._get_split_receivable_op_vals(payment, amounts['amount'], amounts['amount_converted']))
-            account_payment = payment.online_account_payment_id
-            payment_receivable_line = account_payment.move_id.line_ids.filtered(lambda line: line.account_id == account_payment.destination_account_id)
-            online_payment_to_receivable_lines[payment] = split_receivable_line | payment_receivable_line
-
-        data['online_payment_to_receivable_lines'] = online_payment_to_receivable_lines
-        return data
-
-    def _get_split_receivable_op_vals(self, payment, amount, amount_converted):
-        partner = payment.online_account_payment_id.partner_id
-        accounting_partner = self.env["res.partner"]._find_accounting_partner(partner)
-        if not accounting_partner:
-            raise UserError(_("The partner of the POS online payment (id=%d) could not be found", payment.id))
-        partial_vals = {
-            'account_id': accounting_partner.property_account_receivable_id.id,
-            'move_id': self.move_id.id,
-            'partner_id': accounting_partner.id,
-            'name': '%s - %s (%s)' % (self.name, payment.payment_method_id.name, payment.online_account_payment_id.payment_method_line_id.payment_provider_id.name),
-        }
-        return self._debit_amounts(partial_vals, amount, amount_converted)
-
-    def _reconcile_account_move_lines(self, data):
-        data = super()._reconcile_account_move_lines(data)
-        online_payment_to_receivable_lines = data.get('online_payment_to_receivable_lines')
-
-        for payment, lines in online_payment_to_receivable_lines.items():
-            lines.filtered(lambda line: not line.reconciled).reconcile()
-
-        return data
